src/storage/database.rs

```diff
@@ -0,0 +1,108 @@
+use super::Blob;
+use failure::{Error, Fail};
+use postgres::{transaction::Transaction, Connection};
+
+#[derive(Debug, Fail)]
+#[fail(display = "the path is not present in the database")]
+struct PathNotFoundError;
+
+pub(crate) struct DatabaseBackend<'a> {
+    conn: &'a Connection,
+}
+
+impl<'a> DatabaseBackend<'a> {
+    pub(crate) fn new(conn: &'a Connection) -> Self {
+        Self { conn }
+    }
+
+    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+        let rows = self.conn.query(
+            "SELECT path, mime, date_updated, content FROM files WHERE path = $1;",
+            &[&path],
+        )?;
+
+        if rows.is_empty() {
+            Err(PathNotFoundError.into())
+        } else {
+            let row = rows.get(0);
+            Ok(Blob {
+                path: row.get("path"),
+                mime: row.get("mime"),
+                date_updated: row.get("date_updated"),
+                content: row.get("content"),
+            })
+        }
+    }
+    pub(super) fn store_batch(&self, batch: &[Blob], trans: &Transaction) -> Result<(), Error> {
+        for blob in batch {
+            // check if file already exists in database
+            let rows = self
+                .conn
+                .query("SELECT COUNT(*) FROM files WHERE path = $1", &[&blob.path])?;
+
+            if rows.get(0).get::<usize, i64>(0) == 0 {
+                trans.query(
+                    "INSERT INTO files (path, mime, content) VALUES ($1, $2, $3)",
+                    &[&blob.path, &blob.mime, &blob.content],
+                )?;
+            } else {
+                trans.query(
+                    "UPDATE files SET mime = $2, content = $3, date_updated = NOW() \
+                                WHERE path = $1",
+                    &[&blob.path, &blob.mime, &blob.content],
+                )?;
+            }
```

Comment:
All of this can be simplified with a single [`INSERT INTO ... ON CONFLICT SET ...`](https://www.postgresql.org/docs/9.5/sql-insert.html#SQL-ON-CONFLICT) query.